src/engraving/compat/midi/compatmidirenderinternal.cpp

```diff
@@ -304,6 +307,149 @@ static void collectVibrato(int channel,
     pitchWheelRenderer.addPitchWheelFunction(func, channel, staffIdx, effect);
 }
 
+static void addConstPitchWheel(int tick, float value, PitchWheelRenderer& pitchWheelRenderer, int channel, staff_idx_t staffIdx,
+                               MidiInstrumentEffect effect)
+{
+    const float scale = (float)wheelSpec.mLimit / wheelSpec.mAmplitude;
+
+    PitchWheelRenderer::PitchWheelFunction pitchWheelConstFunc;
+    auto constFunc = [value, scale] (uint32_t tick) {
+        UNUSED(tick)
+        return value * scale;
+    };
+
+    pitchWheelConstFunc.func = constFunc;
+    pitchWheelConstFunc.mStartTick = tick;
+    pitchWheelConstFunc.mEndTick = tick + wheelSpec.mStep;
+    pitchWheelRenderer.addPitchWheelFunction(pitchWheelConstFunc, channel, staffIdx, effect);
+}
+
+static bool shouldProceedBend(const Note* note)
+{
+    GuitarBend* bendFor = note->bendFor();
+    Note* baseNote = bendFor->startNoteOfChain();
+
+    GuitarBend* firstBend = baseNote->bendFor();
+    if (firstBend && firstBend->type() == GuitarBendType::PRE_BEND) {
+        Note* nextNote = firstBend->endNote();
+        if (nextNote) {
+            baseNote = nextNote;
+        }
+    }
+
+    while (baseNote->tieFor()) {
+        Tie* tieFor = baseNote->tieFor();
+        Note* nextNote = tieFor->endNote();
+        IF_ASSERT_FAILED(nextNote) {
+            LOGE() << "failed to collect bend for note " << note->pitch() << " on measure " << note->chord()->measure()->no();
+            return false;
+        }
+
+        baseNote = nextNote;
+    }
+
+    return baseNote == note;
+}
+
+static void collectGuitarBend(const Note* note,
+                              int channel,
+                              int onTime, int graceOffset, int previousChordTicks,
+                              PitchWheelRenderer& pitchWheelRenderer, MidiInstrumentEffect effect)
+{
+    if (!shouldProceedBend(note)) {
+        return;
+    }
+
+    int curPitchBendSegmentStart = onTime;
+    int curPitchBendSegmentEnd = 0;
+
+    int quarterOffsetFromStartNote = 0;
+    int currentQuarterTones = 0;
+
+    if (note->bendFor()->type() == GuitarBendType::GRACE_NOTE_BEND) {
+        curPitchBendSegmentStart -= graceOffset;
+    }
+
+    const float scale = (float)wheelSpec.mLimit / wheelSpec.mAmplitude;
+
+    while (note->bendFor() || note->tieFor()) {
+        GuitarBend* bendFor = note->bendFor();
+        int duration = 0;
+        if (bendFor && bendFor->type() == GuitarBendType::GRACE_NOTE_BEND) {
+            duration = (previousChordTicks == -1) ? GRACE_BEND_DURATION : std::min(previousChordTicks / 2, GRACE_BEND_DURATION);
+        } else {
+            duration = note->chord()->actualTicks().ticks();
+        }
+        curPitchBendSegmentEnd = curPitchBendSegmentStart + duration;
+
+        if (bendFor) {
+            Note* endNote = bendFor->endNote();
```

Comment:
Please make all pointers const if the objects they point to should not be modifiable